### PR TITLE
feat: add eip712 sign

### PR DESCRIPTION
**Pull request by Soptq**
This PR adds eip712 sign to the safe-apps-sdk and safe-apps-provider.

### CODE REVIEW

**Comment by Soptq** [comment]
> **CLA Assistant Lite:** Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://gnosis-safe.io/cla/) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> _**I have read the CLA Document and I hereby sign the CLA**_
> 
> You can retrigger the CLA Action by commenting **recheckcla** in this Pull Request

I have read the CLA Document and I hereby sign the CLA

**Comment by Soptq** [comment]
I have read the CLA Document and I hereby sign the CLA

File: packages/safe-apps-sdk/src/types/sdk.ts [hunk 1]
```diff
@@ -32,6 +32,19 @@ export type SignMessageParams = {
...
 
+interface Type {
+  name: string;
+  type: string;
+}
+
+export type TypedMessageTypes = { [key: string]: Type[] };
+
+export type SignTypedMessageParams = {
+  domain: any;
```
Comment (katspaugh) [inline packages/safe-apps-sdk/src/types/sdk.ts:43]: Should this be a string?

File: packages/safe-apps-sdk/src/types/sdk.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { ChainInfo as _ChainInfo } from '@gnosis.pm/safe-react-gateway-sdk';
 export type ChainInfo = Pick<
   _ChainInfo,
   'chainName' | 'chainId' | 'shortName' | 'nativeCurrency' | 'blockExplorerUriTemplate'
->;
+  >;
```
Comment (katspaugh) [inline packages/safe-apps-sdk/src/types/sdk.ts:6]: Revert this pls.

File: packages/safe-apps-sdk/src/types/sdk.ts (hunk 1 above)
Comment (Soptq) [reply packages/safe-apps-sdk/src/types/sdk.ts:43]: To my best knowledge, `domain` can be anything. For example in https://medium.com/metamask/eip712-is-coming-what-to-expect-and-how-to-use-it-bb92fd1a7a26:

```
const domainData = {
    name: "My amazing dApp",
    version: "2",
    chainId: parseInt(web3.version.network, 10),
    verifyingContract: "0x1C56346CD2A2Bf3202F771f50d3D14a367B48070",
    salt: "0xf2d857f4a3edcb9b78b4d503bfe733db1e3f6cdc2b7971ee739626c97e86a558"
};
```

File: packages/safe-apps-sdk/src/types/sdk.ts (hunk 2 above)
Comment (Soptq) [reply packages/safe-apps-sdk/src/types/sdk.ts]: Done. Thanks for the notifying~

File: packages/safe-apps-sdk/src/types/sdk.ts (hunk 1 above)
Comment (katspaugh) [reply packages/safe-apps-sdk/src/types/sdk.ts:43]: Pretty sure you can't pass just any JS value. E.g. surely not HTML elements or functions?
How about `Record<any, any> | Record<any, any>[]`?

Comment (Soptq) [reply packages/safe-apps-sdk/src/types/sdk.ts]: Done, changed the type of `domain` from `any` to `Record<string, any>` (I think `domain` can't contain array either)

File: packages/safe-apps-sdk/dist/src/types/sdk.d.ts [hunk 3]
```diff
@@ -22,6 +22,18 @@ export declare type GetBalanceParams = {
 export declare type SignMessageParams = {
     message: string;
 };
+interface TypedDataTypes {
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/dist/src/types/sdk.d.ts:25]:
```
export interface TypedDataDomain {
    name?: string;
    version?: string;
    chainId?: BigNumberish;
    verifyingContract?: string;
    salt?: BytesLike;
};

export interface TypedDataField {
    name: string;
    type: string;
};

export type TypedDataTypes = Record<string, Array<TypedDataField>>
```

taken from https://github.com/ethers-io/ethers.js/blob/608864fc3f00390e1260048a157af00378a98e41/packages/abstract-signer/src.ts/index.ts#L25

File: packages/safe-apps-sdk/dist/src/types/sdk.d.ts [hunk 4]
```diff
@@ -22,6 +22,18 @@ export declare type GetBalanceParams = {
...
+    name: string;
+    type: string;
+}
+export declare type TypedMessageTypes = {
+    [key: string]: TypedDataTypes[];
+};
+export declare type SignTypedMessageParams = {
+    domain: Record<string, any>;
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/dist/src/types/sdk.d.ts:33]: I think we can add a more specific type for the domain

```
export interface TypedDataDomain {
    name?: string;
    version?: string;
    chainId?: BigNumberish;
    verifyingContract?: string;
    salt?: BytesLike;
};
```

taken from https://github.com/ethers-io/ethers.js/blob/608864fc3f00390e1260048a157af00378a98e41/packages/abstract-signer/src.ts/index.ts#L25

The description of the fields can be found in the [EIP-712 Standard](https://eips.ethereum.org/EIPS/eip-712) in the `Definition of domainSeparator` section

**Comment by mikhailxyz** [comment]
Could you please also add a changeset for your changes? You can do that by running `npx changeset`

File: packages/safe-apps-provider/src/provider.ts [hunk 5]
```diff
@@ -64,6 +64,18 @@ export class SafeAppProvider extends EventEmitter implements EIP1193Provider {
         return '0x';
       }
 
+      case 'eth_signTypedData_v4': {
```
Comment (mikhailxyz) [inline packages/safe-apps-provider/src/provider.ts:67]: let's also add support for `eth_signTypedData`, version postfix is metamask specific and not a part of the official eip

File: packages/safe-apps-provider/src/provider.ts [hunk 6]
```diff
@@ -64,6 +64,18 @@ export class SafeAppProvider extends EventEmitter implements EIP1193Provider {
...
+        const [message, address] = params;
+
+        if (this.safe.safeAddress.toLowerCase() !== address.toLowerCase()) {
+          throw new Error('The address or message hash is invalid');
```
[the remark this answers is not in the document]
Comment (Soptq) [reply packages/safe-apps-provider/src/provider.ts]: Updated, thanks!

File: packages/safe-apps-sdk/dist/src/types/sdk.d.ts (hunk 4 above)
Comment (Soptq) [reply packages/safe-apps-sdk/dist/src/types/sdk.d.ts]: Updated, thanks!

**Comment by Soptq** [comment]
> Could you please also add a changeset for your changes? You can do that by running `npx changeset`

Added ;)

**Review by mikhailxyz — APPROVED** [review]

**Comment by ennisjo** [comment]
@Soptq - can you ping my telegram @ennis_safe to organise payment?

**Pull request merged**